libcudacxx/include/cuda/std/detail/libcxx/include/__algorithm/copy.h

```diff
@@ -83,9 +84,19 @@ __copy(_Tp* __first, _Tp* __last, _Up* __result)
     {
       return {__last, __result + __n};
     }
-    for (ptrdiff_t __i = 0; __i < __n; ++__i)
+    if (__result - __first > 0)
     {
-      *(__result + __i) = *(__first + __i);
+      for (ptrdiff_t __i = __n; __i > 0; --__i)
+      {
+        *(__result + __i - 1) = *(__first + __i - 1);
+      }
+    }
+    else
+    {
+      for (ptrdiff_t __i = 0; __i < __n; ++__i)
+      {
+        *(__result + __i) = *(__first + __i);
+      }
     }
```

Comment:
I am not sure what to think about this. I would rather have the invalid usecases switch to `copy_backwards` in the right cases

Reply:
"Overlap of ranges is formally permitted, but leads to unpredictable ordering of the results." (ref. cppreference/copy_n)

I think it's fine for us to actually strengthen the guarantee here. It'd be slightly more awkward to deal with replacing `copy_n` with `copy_backward` instead of just fixing the behavior in one place.